### PR TITLE
numpy v2 compatibility

### PR DESCRIPTION
**Pull request by megies**
The current main branch is not ready for numpy v2.

Aside from some rather trivial namespace changes, the main change to be aware of is [changes to up-/downcasting rules](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#changes-to-numpy-data-type-promotion).

The steim2 roundtrip test ran into one of these issues. The simplemseed code internally does computations on a mix of `uint32` and `int32` and numpy (both v1 and v2) is upcasting the result to `int64`. 64 bit items like that are internally then used in assignments into 32 bit arrays. In numpy v1, such assignments that would overflow were silently working and overflowing into negative numbers. In numpy v2, such an assignment that would overflow actually does exactly that and raise an `OverflowError`.

Long story short, @crotwell, my guess is that the test suite more than likely does not cover all potential cases of such overflow scenarios. The code should ideally be checked for all occasions where `uint32` and `int32` (or `uint32` and native Python integers etc etc) are used in combined computations and either these should be changed so that all members of a given computation share the same type or the result needs to be explicitly cast to the correct type.

I think the diff illustrates the problem. I wasn't entirely sure which way to "fix" the issue was best, since I'm also not well versed with bitwise operations and potential implications or the possible reasoning behind choosing `uint` types in some of the parts of the code.

### CODE REVIEW

**Comment by megies** [comment]
### numpy v1
Screenshot from 2024-07-22 16-46-49


---
### numpy v2
Screenshot from 2024-07-22 16-43-17

**Comment by megies** [comment]
### numpy v1
Screenshot from 2024-07-22 16-48-24
---
### numpy v2
Screenshot from 2024-07-22 16-48-02

**Comment by crotwell** [comment]
OK, will look into this over next couple days.

**Comment by crotwell** [comment]
@megies I have made changes to the code so no more numpy 1.2->2.0 warnings from `np._set_promotion_state("weak_and_warn")` and tests pass both under numpy1.26.4 and numpy2.0.0. 
Hopefully address this, but grateful if you can test to.

Very much appreciate your feedback and insight on this.

**Comment by megies** [comment]
@crotwell changes are looking nice to me, not the massive expert on the numpy change either but I think fixing these types more should be much safer.

I wanna check here again, looks like still some mixed types, will look into it:
https://github.com/crotwell/simplemseed/commit/08f3f4caa24c25475564ec2f8d9ce84fb88ff656#diff-ff7d67a481970bff060f402647e1d9152571c07361c5cc5fbbaa62a5e691e89dL508

Edit:
That part as expected still ends up in a return value of `np.int64` (due to the mix of `uint32` and `int32`) which I understand is probably unwanted, IIUC.

```
tests/test_mseed3.py ...
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/megies/git/simplemseed/src/simplemseed/steim2.py(513)steimPackWord()
-> return val
(Pdb) l
508  	        val = np.bitwise_or(np.left_shift(val, np.uint32(nbits)) , np.bitwise_and(np.int32(diff[i]), np.uint32(bitmask)))
509  	        i += 1
510  	    if submask != 0:
511  	        val = np.bitwise_or(val, np.left_shift(np.uint32(submask), np.uint32(30)))
512  	    breakpoint()
513  ->	    return val
[EOF]
(Pdb) type(val)
<class 'numpy.int64'>
```

I tried to find if numpy can be told not to upcast, but I can't find anything. In any case this one does not throw an exception anymore since it looks like you changed the type of the `SteimFrame` to `uint32`, so it seems in the test case it works because the value that gets assigned to the frame is of `np.int64` and positive, not sure if the values you produce could end up being in the negative range of `np.int64` which then would probably error out again when trying to set the value on the frame.

**Comment by crotwell** [comment]
This is tricky, I think a numpy operation on a int32 and a uint32 always gives int64 even in cases where it can't actually overflow like bitwise AND.

Added some more casting, checks and tests. Also now encoding is argument, so can do

```
stream.write(filename, format="MSEED3", encoding="STEIM2")
```
